### PR TITLE
[Metrics] Detach bugfix

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/4098
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/4266
Make sure that the metrics states gets detached from the computational graph to prevent memory consumption from growing.
Also make sure that results gets detached for logging.
Also added note on how metrics can be backproped.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by teddykoker — APPROVED** [review]
Good catch nice work :)

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -144,6 +144,18 @@ Example implementation:
         def compute(self):
             return self.correct.float() / self.total
 
+Metrics support backpropergation, if all computations involved in the metric calculation
```
Comment (nathanpainchaud) [inline docs/source/metrics.rst:147]: 'backpropergation' -> 'backpropagation'

File: docs/source/metrics.rst [hunk 2]
```diff
@@ -144,6 +144,18 @@ Example implementation:
...
+are differentiable. However, note that the cashed state is detached from the computational
+graph and cannot be backpropergated. Not doing this would mean storing the computational
+graph for each update call, which can lead to out-of-memory errors.
+In practise this means that:
+
+.. code-block:: python
+    metric = MyMetric()
+    val = metric(pred, target) # this value can be backpropergated
+    val = metric.compute() # this value cannot be backpropergated
```
Comment (nathanpainchaud) [inline docs/source/metrics.rst:156]: 'backpropergated' -> 'backpropagated'

**Review by nathanpainchaud — COMMENTED** [review]
Just a few typos in the added documentation.

File: docs/source/metrics.rst (hunk 1 above)
Comment (SkafteNicki) [reply docs/source/metrics.rst]: Thanks 👍

File: docs/source/metrics.rst (hunk 2 above)
Comment (SkafteNicki) [reply docs/source/metrics.rst]: Thanks 👍

File: pytorch_lightning/metrics/metric.py [hunk 3]
```diff
@@ -142,7 +142,8 @@ def forward(self, *args, **kwargs):
         Automatically calls ``update()``. Returns the metric value over inputs if ``compute_on_step`` is True.
         """
         # add current step
-        self.update(*args, **kwargs)
+        with torch.no_grad():
```
Comment (ananyahjha93) [inline pytorch_lightning/metrics/metric.py:145]: @teddykoker @SkafteNicki  also when we are using dist_sync_on_step as True, that should be a no_grad op. Since DDP will expect grads to be computed only for this GPU and then sync those when backpropagating.

So https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/distributed.py#L76 needs to be done in no_grad if I am not wrong.

**Review by ananyahjha93 — CHANGES_REQUESTED** [review]

File: docs/source/metrics.rst [hunk 4]
```diff
@@ -144,6 +144,18 @@ Example implementation:
         def compute(self):
             return self.correct.float() / self.total
 
+Metrics support backpropagation, if all computations involved in the metric calculation
+are differentiable. However, note that the cashed state is detached from the computational
+graph and cannot be backpropagated. Not doing this would mean storing the computational
+graph for each update call, which can lead to out-of-memory errors.
+In practise this means that:
+
+.. code-block:: python
```
Comment (teddykoker) [inline docs/source/metrics.rst:153]: `
```suggestion
.. code-block:: python

```

Comment (teddykoker) [reply docs/source/metrics.rst:153]: Need new line for docs

File: pytorch_lightning/metrics/metric.py (hunk 3 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:145]: If we do that will it not break backpropagation, as everything gets detached from the computational graph?
We should probably be able to test this.

**Review by nateraw — COMMENTED** [review]
this is a huge catch, thank you @SkafteNicki ...I was goin OOM on GPU when training on large dataset last night, and this is definitely the culprit.

File: pytorch_lightning/metrics/metric.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py:153]: I don't think this should be `no_grad` for the exact same reason @SkafteNicki mentioned. If you backdrop, the device-changes (as well as dip sync) need to be recorded by the graph to make sure you can go back to the original device during backdrop (and to also find all the activations there).

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by ananyahjha93 — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

**Comment by SeanNaren** [comment]
Looks good! I was contemplating if the `detach()` should move within the `compute()` function, but then realised there maybe be a case where you wouldn't want to detach computed metrics.
